lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -136,22 +129,24 @@ def check_assoc(assoc)
         def add_resource_relationships(attrs, serializer, options = {})
           options[:add_included] = options.fetch(:add_included, true)
 
+          attrs[:relationships] ||= {} if serializer.associations.any?
           serializer.associations.each do |association|
             key = association.key
             serializer = association.serializer
             opts = association.options
-
-            attrs[:relationships] ||= {}
-
-            if serializer.respond_to?(:each)
-              add_relationships(attrs, key, serializer)
-            else
-              if opts[:virtual_value]
-                add_relationship(attrs, key, nil, opts[:virtual_value])
-              else
-                add_relationship(attrs, key, serializer)
-              end
-            end
+            value = if serializer.respond_to?(:each)
+                      serializer.map { |s| resource_identifier(s) }
+                    else
+                      if opts[:virtual_value]
+                        opts[:virtual_value]
+                      elsif serializer && serializer.object
+                        resource_identifier(serializer)
+                      else
+                        nil
+                      end
+                    end
```

Comment:
yeah, definitely a latent object in here for serializer and serializer collections...  it would also make naming blocks easier e.g. `serializers.each { |serializer| }`